docs_src/design/adr/device-service/0021-Profile-Changes.md

```diff
@@ -0,0 +1,126 @@
+
+### Association Changes
+
+Today, a device profile is associated to a device via its name (the ProfileName on the Device struct).  The profile can also associated to events and readings (via the ProfileName in the Event and Reading objects).
+
+Profiles can be associated to more than one device.  Profiles can be associated to multiple events/readings.
+
+Rules about change of a device profile association have never been explicitly defined.  While the section above outline what can/cannot be changed in the profile and under what circumstances, this section outlines whether a device, event or reading can alter the associated device profile and under what circumstances.
+
+Change of the device profile on a device – never allowed unless there are no events or readings in the system.  There is too much risk that an existing event or reading also contains the device profile reference – causing complexity in knowing what information is in the event/reading.
+
+A provision watcher also has an association to a device profile.  However, the device profile is only used to know the type of device to create in case of provisioning.  Therefore, if a device profile changes, it does not impact the provision watcher.
+
+A device profile cannot be deleted while it is associated to a provision watcher
+
+Change of the device profile on an event or reading is never allowed.  Events and readings are immutable.
+
+### Applicability
+
+This document dictates the rules around allowed change to device profiles (and associations).  All EdgeX services will abide by these rules (implementing appropriate checks and validation as necessary to enforce the rules) when this ADR is approved to include:
+- service APIs
+- user interfaces (like the Edge GUI)
+- client tools (like the EdgeX CLI)
+
+## Proposed Design
+
+The following metadata API changes are suggested as a means to implement the device profile rules of this ADR
+
+- Block Device DELETE when there is any associated event/reading exists.
+- Block Profile (Upload) PUT when there is any associated device/event/reading exists.
+- Block Profile DELETE when there is any associated device/event/reading exists.
+- Allow empty Profile POST (containing no device resources or commands)
+- Add Profile General Property PATCH API
+- Add Profile Device Resource POST API
+- Add Profile Device Resource PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Resource DELETE API (check it's not associated to any device, reading or event)
+- Add Profile Device Command POST API
+- Add Profile Device Command PATCH API (allow to modify Description and IsHidden only)
+- Add Profile Device Command DELETE API (check it's not associated to any device, reading or event)
+
+*Note* - to be determined: how to allow for device resource or device command changes (outside of Description and isHidden) when the device profile is not yet associated to any object
```

Comment:
I'm not sure what this means?

Reply:
Should be allowed.

Reply:
This note is based on earlier exchange in this PR between @cloudxxx8 and I.  Cloud had provided for no means to change the device resources or device commands.  I feel (and I think @lenny-intel is agreeing) there should be (given that the profile is not yet associated to any object).  Cloud - any reason why we shouldn't be able to change the resources or commands?

Reply:
If the Device Profile is associated to any existing Devices, Resource or Device Command not be changed.
However, modifying Description or IsHidden should be fine.

Reply:
> If the Device Profile is associated to any existing Devices, Resource or Device Command not be changed. However, modifying Description or IsHidden should be fine.

@cloudxxx8 , we are say the opposite, i.e. not associated to any existing Devices, Resource or Device Command, so should allow change.

Reply:
I see, you are right
-> not associated to any existing Devices, Resource or Device Command, so should allow change.